src/main/kotlin/io/meshcloud/dockerosb/persistence/GitHandlerService.kt

```diff
@@ -0,0 +1,321 @@
+      return
+    }
+
+    if (!hasLocalCommits.get()) {
+      // note: we do also not execute a fetch in this case because if the unipipe-osb is just idling
+      // there's no sense in us fetching from the remote all the time. Consumers can explicitly call
+      // pullFastForwardOnly if they want an up to date copy
+      log.info { "synchronizeWithRemoteRepository called, but no new local commits found - skipping." }
+      return
+    }
+
+    log.info { "Starting synchronizeWithRemoteRepository" }
+
+    try {
+      val pushedSuccessfully = fetchMergePush()
+      if (pushedSuccessfully) {
+        hasLocalCommits.set(false)
+      }
+
+      log.info { "Completed synchronizeWithRemoteRepository" }
+    } catch (ex: Exception) {
+      log.error(ex) { "Failed synchronizeWithRemoteRepository" }
+    }
+  }
+
+  private fun fetchMergePush(): Boolean {
+    // fetch from the remote
+    git.fetch()
+        .apply { remote = "origin" }
+        .call()
+
+    // merge changes - this happens locally in our repository
+    val merge = git.merge()
+        .include(git.repository.resolve("origin/${gitConfig.remoteBranch}"))
+        .apply {
+          setFastForward(MergeCommand.FastForwardMode.FF) // fast-forward if possible, otherwise merge
+          setCommit(true)
+          setMessage("OSB API: auto-merging remote changes")
+
+          // unfortunately jgit does not support "git merge --recursive -X ours"... yet but it will
+          // https://github.com/eclipse/jgit/commit/8210f29fe43ccd35e7d2ed3ed45a84a75b2717c4
+          setStrategy(MergeStrategy.RECURSIVE) // we are
+        }
+        .call()
+
+    when (merge.mergeStatus) {
+      MergeResult.MergeStatus.CONFLICTING -> {
+        val files = merge.conflicts.entries.map { it.key }
+        resolveMergeConflictsUsingOurs(files)
+      }
+      MergeResult.MergeStatus.CHECKOUT_CONFLICT,
+      MergeResult.MergeStatus.FAILED -> {
+        log.warn { "Merge failed with status ${merge.mergeStatus.name}. Will retry on next periodic sync." }
+        logFailedMergeDetails(merge)
+        recoverFromFailedMerge()
+      }
+
+      MergeResult.MergeStatus.ALREADY_UP_TO_DATE,
+      MergeResult.MergeStatus.FAST_FORWARD,
+      MergeResult.MergeStatus.FAST_FORWARD_SQUASHED,
+      MergeResult.MergeStatus.MERGED,
+      MergeResult.MergeStatus.MERGED_NOT_COMMITTED,
+      MergeResult.MergeStatus.MERGED_SQUASHED,
+      MergeResult.MergeStatus.MERGED_SQUASHED_NOT_COMMITTED -> {
+        log.info { "Merge succeeded with status ${merge.mergeStatus.name}." }
+      }
+
+      MergeResult.MergeStatus.ABORTED,
+      MergeResult.MergeStatus.NOT_SUPPORTED -> {
+        log.error { "Merge failed with unexpected status ${merge.mergeStatus.name}. Taking no further action. An operator needs to resolve conflicts on the remote repository." }
+        recoverFromFailedMerge()
+      }
+      null -> throw KotlinNullPointerException("merge.mergeStatus")
```

Comment:
yes, that's just a java -> kotlin null quirk. I wanted to make the switch statemement exhaustive without using a default/else block because it forces us to handle every error code explicitly. the null handling is then required to make  kotlinc happy

Reply:
I've converted it to an ISE, no preference on what the ideal exception type would be (I just think kotlin throws NPEs too if a `!!` turns out to be false